scripts/usb_scan.py

```diff
@@ -0,0 +1,725 @@
+
+    @abc.abstractmethod
+    def is_class_hub(self):
+        """ check if this belongs to a hub
+
+        :return: bool, if this belongs to a hub
+        """
+        pass
+
+    @abc.abstractmethod
+    def is_child_of(self, parent):
+        """ check if this is a child of parent
+
+        :param parent:(UsbObject) the parent to check against
+        :return:
+        """
+        pass
+
+    @staticmethod
+    def validate_int(s, base=10):
+        """ validate if a string can be converted to int
+
+        :param s:(str) the string to be converted
+        :param base:(int) the radix base of integer to convect
+        :return:(bool) if conversion is OK
+        """
+        try:
+            int(s, base)
+            return True
+        except (TypeError, ValueError):
+            return False
+
+
+class UsbDevice(UsbObject):
+    """ Class for USB device, save USB properties in UsbObject dict
+
+    interfaces:([UsbInterface]) list of USB interfaces belonging to this device
+    allow:(bool) if passthrough is permitted
+    """
+    _DESC_VENDOR = "ID_VENDOR_FROM_DATABASE"
+    _DESC_PRODUCT = "ID_MODEL_FROM_DATABASE"
+
+    _VERSION = "version"
+    _ID_VENDOR = "idVendor"
+    _ID_PRODUCT = "idProduct"
+    _BCD_DEVICE = "bcdDevice"
+    _SERIAL = "serial"
+    _CLASS = "bDeviceClass"
+    _CONF_VALUE = "bConfigurationValue"
+    _NUM_INTERFACES = "bNumInterfaces"
+
+    _PROPS_1 = [_DESC_VENDOR, _DESC_PRODUCT]
+    _PROPS_2 = [_VERSION, _ID_VENDOR, _ID_PRODUCT, _BCD_DEVICE, _SERIAL, _CLASS, _CONF_VALUE, _NUM_INTERFACES]
+    _PROPS = _PROPS_1 + _PROPS_2
+    _PROPS_NONABLE = _PROPS_1 + [_SERIAL]
+
+    def __init__(self, node, props1, props2):
+        """ initialise UsbDevice, set node and properties
+
+        :param node(str): device node
+        :param props1(pyudev.Device): device, to get properties from UDEV database
+        :param props2(pyudev.Device.attributes): device attributes, to get properties from sysfs
+        """
+        super(UsbDevice, self).__init__(node)
+
+        for p in self._PROPS_1:
+            if props1.get(p) is not None:
+                self[p] = props1.get(p)
+        for p in self._PROPS_2:
+            if props2.get(p) is not None:
+                self[p] = props2.get(p)
+        for p in self._PROPS_NONABLE:
+            if p not in self:
+                self[p] = ""
+
+        self.interfaces = set()
+        self.allow = False
+
+    def debug_str(self, level=0):
+        s = super(UsbDevice, self).debug_str(level)
+        for i in self.interfaces:
+            s += i.debug_str(level + 1)
+        s += self.indent(level + 1) + "Allow: " + str(self.allow)
+        # s += self.indent(level + 1) + "Storages: " + str(self.get_all_storages())
+        return s
+
+    def is_initialized(self):
+        for p in self._PROPS:
+            if p not in self:
+                return False
+        # usb_descriptor_attr_le16(idVendor, "%04x\n");
```

Comment:
If this code is unused please remove it.

Reply:
OK.